awx/main/tasks.py

```diff
@@ -660,6 +666,43 @@ def update_host_smart_inventory_memberships():
         smart_inventory.update_computed_fields()
 
 
+@task()
+def migrate_legacy_event_data(tblname):
+    if 'event' not in tblname:
+        return
+    with advisory_lock(f'bigint_migration_{tblname}', wait=False) as acquired:
+        if acquired is False:
+            return
+        chunk = 100000
+
+        def _remaining():
+            try:
+                cursor.execute(f'SELECT MAX(id) FROM _old_{tblname};')
+                return cursor.fetchone()[0]
+            except ProgrammingError:
+                # the table is gone (migration is unnecessary)
+                return None
+
+        with connection.cursor() as cursor:
+            total_rows = _remaining()
+            while total_rows is not None:
+                with transaction.atomic():
+                    cursor.execute(
+                        f'INSERT INTO {tblname} SELECT * FROM _old_{tblname} ORDER BY id DESC LIMIT {chunk};'
```

Comment:
Say that 1 year from now we introduce a new field to the JobEvent model that did not previously exist. Or say we delete a field that wasn't being used. That'll cause a mismatch between these two tables, right?

Reply:
Yes, though the idea is that you only one this *once*, and this it would finish within a few hours - you wouldn't enter this code path once the `_old_` tables are empty/cease to exist.

Reply:
Does this imply that *all* upgrades must go through a 3.7.x, and cannot skip over it?

Reply:
... and that if we have a later migration for 3.7 that touches the JobEvent model, things are going to go poorly?

Reply:
If a migration on one of these tables that affects the _column count_ comes *after* this in 3.7, that could cause issues, yes.

Reply:
> Does this imply that all upgrades must go through a 3.7.x, and cannot skip over it?

@wenottingham not, if you went from e.g., 3.6 ---> 3.8, it's still going to follow these migrations sequentially.

Reply:
If we wanted to be *very* careful about this, I could explicitly itemize the columns instead of using `SELECT *` here (essentially, freeze the current column set in time).

Reply:
The columns in the _old_ table _are_ frozen in time, because the old table gets created when the schema is in its current state.  The `SELECT *` is getting its column list from the old table. 
 So I believe this insert would still work even if you added a column later (as long as the new column isn't a required field).

If you removed or renamed a column then the new migration would need to make sure the table data copy had completed before making the schema change.  And in this case the whole data copy would have to run during the upgrade.  But this would not be too hard to do at the time of writing the new migration.

Reply:
> if you went from e.g., 3.6 ---> 3.8, it's still going to follow these migrations sequentially.

But those 3.8/3.9/etc migrations would run while this copy is going on in the background, correct?

Reply:
Yes, or perhaps before it starts.  But as long as the new migrations don't introduce changes that break the `INSERT` statement, it seems to me that this should work fine.

Reply:
> So I believe this insert would still work even if you added a column later (as long as the new column isn't a required field).

So the more I think about this (and I haven't actually *tested* it), I suspect that this would work just fine so long as:

- none of the _existing_ column *types* were changed in a way that caused inserts to fail
- none of the _existing_ columns were *removed* in such a way that the `_old_` table now had columns which didn't exist in the *new* table schema.

Reply:
@wenottingham 

> But those 3.8/3.9/etc migrations would run while this copy is going on in the background, correct?

On upgrade, no.  All of the migrations would _finish_, then AWX services (including the dispatcher would _start_, and **then**, the data migration would commence in the background.